src/v/cluster/producer_state.h

```diff
@@ -0,0 +1,208 @@
+/*
+ * Copyright 2023 Redpanda Data, Inc.
+ *
+ * Use of this software is governed by the Business Source License
+ * included in the file licenses/BSL.md
+ *
+ * As of the Change Date specified in that file, in accordance with
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#pragma once
+
+#include "cluster/types.h"
+#include "model/record.h"
+#include "utils/intrusive_list_helpers.h"
+#include "utils/mutex.h"
+#include "utils/rwlock.h"
+
+#include <seastar/core/shared_future.hh>
+#include <seastar/util/defer.hh>
+
+// Befriended to expose internal state in tests.
+struct test_fixture;
+
+namespace cluster {
+
+template<class Func>
+concept AcceptsUnits = requires(Func f, ssx::semaphore_units units) {
+    f(std::move(units));
+};
+
+class producer_state_manager;
+class producer_state;
+class request;
+
+using producer_ptr = ss::lw_shared_ptr<producer_state>;
+// Note: A shared_promise doesn't guarantee available() to be true
+// right after set_value(), this is an implementation quirk, be
+// mindful of that behavior when using it. We have a test for
+// it in expiring_promise_test
+using result_promise_t = ss::shared_promise<result<kafka_result>>;
+using request_ptr = ss::lw_shared_ptr<request>;
+using seq_t = int32_t;
+
+/// A request for a given sequence range, both inclusive.
+/// The sequence numbers are stamped by the client and are a part
+/// of batch header. A request can either be in progress or completed
+/// depending on the whether the holding promise is set.
+class request {
+public:
+    explicit request(
+      seq_t first, seq_t last, model::term_id term, result_promise_t res)
+      : _first_sequence(first)
+      , _last_sequence(last)
+      , _term(term)
+      , result(std::move(res)) {}
+
+    template<class ValueType>
+    void set_value(ValueType&& value) {
+        result.set_value(std::forward<ValueType>(value));
+    }
+
+private:
+    seq_t _first_sequence;
+    seq_t _last_sequence;
+    // term in which the request was submitted.
+    model::term_id _term;
+    // Set when the result for this request is finished. This is a shared
+    // promise because a client can retry an already in progress request
+    // (eg: timeouts) and we just chain the retried request request with
+    // with the future from an already in-flight promise with the sequence
+    // number match.
+    result_promise_t result;
+
+    bool in_progress() { return !result.available(); }
+    friend class requests;
+    friend class producer_state;
+};
+
+// A cached buffer of requests, the requests can be in progress / finished.
+// A request is promoted from inflight to finished once it is applied in the
+// log.
+//
+// We retain a maximum of `requests_cached_max` finished requests.
+// Kafka clients only issue requests in batches of 5, the queue is fairly small
+// at all times.
+class requests {
+public:
+    using request_result_t
```

Comment:
why not `checked<request_ptr, model::term_id, errc>`? a function returning three different types of result depending on the context doesn't feel right

Reply:
I meant `checked<request_ptr, errc>`. A function returning a value of type A, type B and type C is confusing and it is often an indicator of a leaking abstraction. An it looks like it's happening there.

We have `new_request(...)` api which either return an error (that's fine), a new request (that's fine) or a projection of an existing request (that's wrong). `new_request(...)` returning existing request is misleading.

I'm proposing to have two methods:
- `get_request()` returning `optional<request_ptr>` (also no need to hide the result field)
- `new_request()` returning `checked<request_ptr, errc>`

This way the api becomes cleaner and we avoid a method returning value of different types depending on the context.

Reply:
>I'm proposing to have two methods:
> get_request() returning optional<request_ptr> (also no need to hide the result field)
> new_request() returning checked<request_ptr, errc>

IMO this is not any simpler because the caller code is complicated and has to know whether there is cached request and if not make another call for a new request. variant kind of encapsulates all of this. We do this pattern in multiple places in the code

Reply:
functionally it's the same but it conveys meaning better:
- the names reflect what the methods are doing
- each method has a single responsibility
- the signature doesn't require an alias (we spend a lot of time on github reviewing PRs where you can't easily jump to declaration so the less indirections we have - the better)

Reply:
Got rid of variant, introduced a state enum.